plugins/in_dbus/in_dbus.c

```diff
@@ -0,0 +1,527 @@
+    }
+    return 0;
+}
+
+dbus_bool_t in_dbus_add_watch(DBusWatch* watch, void* data)
+{
+    struct in_dbus_event* event = data;
+    const unsigned flags = dbus_watch_get_flags(watch);
+    const int fd = dbus_watch_get_unix_fd(watch);
+    unsigned mk_flags = 0;
+    if (flags & DBUS_WATCH_READABLE) {
+        mk_flags |= MK_EVENT_READ;
+    }
+    if (flags & DBUS_WATCH_WRITABLE) {
+        mk_flags |= MK_EVENT_WRITE;
+    }
+    event->watch = watch;
+    mk_event_add(event->evl, fd, FLB_ENGINE_EV_CUSTOM, mk_flags, data);
+    return true;
+}
+
+void in_dbus_remove_watch(DBusWatch* watch, void* data)
+{
+    struct in_dbus_event* event = data;
+    mk_event_del(event->evl, data);
+}
+
+/* cb_collect callback
+ * This callback does very little work, because data is being accumulated
+ * into the messagepack buffer by a separate task. */
+static int in_dbus_collect(struct flb_input_instance *i_ins,
+                           struct flb_config *config, void *in_context)
+{
+    struct flb_in_dbus_config *dbus_config = in_context;
+    if (dbus_config->mp_sbuf.size != 0) {
+        flb_input_chunk_append_raw(i_ins, NULL, 0,
+                                   dbus_config->mp_sbuf.data,
+                                   dbus_config->mp_sbuf.size);
+        msgpack_sbuffer_clear(&dbus_config->mp_sbuf);
+    }
+
+    return 0;
+}
+
+static void in_dbus_reply_error(DBusMessage* msg, DBusConnection* conn,
+                                const char* message)
+{
+    DBusMessage* reply;
+    dbus_uint32_t serial = 0;
+
+    reply = dbus_message_new_error(msg, DBUS_ERROR_FAILED, message);
+    if (dbus_connection_send(conn, reply, &serial)) {
+        dbus_connection_flush(conn);
+    }
+    else {
+        flb_error("[in_dbus] Failed to send error reply");
+    }
+    dbus_message_unref(reply);
+}
+
+static void in_dbus_introspect(struct flb_in_dbus_config *dbus_config,
+                               DBusMessage* msg, DBusConnection* conn)
+{
+    DBusMessage* reply;
+    DBusMessageIter args;
+    dbus_uint32_t serial = 0;
+    const char* description = ""
+        "<!DOCTYPE node PUBLIC \"-//freedesktop//DTD D-BUS Object Introspection 1.0//EN\""
+        "        \"http://www.freedesktop.org/standards/dbus/1.0/introspect.dtd\">"
+        "<node>"
+        "    <interface name=\"com.fluent.fluentbit\">"
+        "        <method name=\"LogData\">"
+        "            <arg name=\"data\" type=\"a{sv}\" direction=\"in\"/>"
+        "            <annotation name=\"org.qtproject.QtDBus.QtTypeName.In0\" value=\"QVariantMap\"/>"
+        "        </method>"
+        "        <method name=\"LogTimestampedData\">"
+        "            <arg name=\"tv_sec\"  type=\"x\" direction=\"in\"/>"
+        "            <arg name=\"tv_nsec\" type=\"x\" direction=\"in\"/>"
+        "            <arg name=\"data\" type=\"a{sv}\" direction=\"in\"/>"
+        "            <annotation name=\"org.qtproject.QtDBus.QtTypeName.In2\" value=\"QVariantMap\"/>"
```

Comment:
Can we remove these annotations? I thought these were just for `qdbusxml2cpp`'s benefit.